docker/tarfile/src.go

```diff
@@ -35,14 +38,63 @@ type layerInfo struct {
 	size int64
 }
 
-// NewSource returns a tarfile.Source for the specified path.
-func NewSource(path string) *Source {
+// NewSource returns a tarfile.Source for the specified path or inputStream.
+// Cannot set path and inputStream at the same time.
+// inputStream is mainly used for the docker-daemon transport.
+func NewSource(path string, inputStream io.ReadCloser) (*Source, error) {
 	// TODO: We could add support for multiple images in a single archive, so
 	//       that people could use docker-archive:opensuse.tar:opensuse:leap as
 	//       the source of an image.
+
+	if path != "" && inputStream != nil {
+		return nil, errors.Errorf("Cannot set both path and inputStream")
+	}
+
+	// Check if path is a gzip file and read its contents if it is
+	isGzip := false
+	if path != "" {
```

Comment:
(This is another argument for splitting `NewSource`: it would be awkward if `NewSource` accepted compressed input for files but not for streams; with two separate constructors it’s more acceptable.)

Reply:
Fixed